src/Components/Web/src/Forms/InputRadioContext.cs

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Components.Forms
+{
+    /// <summary>
+    /// Describes context for an <see cref="InputRadio{TValue}"/> component.
+    /// </summary>
+    public class InputRadioContext
+    {
+        private readonly InputRadioContext? _parentContext;
+
+        /// <summary>
+        /// Gets the name of the input radio group.
+        /// </summary>
+        public string GroupName { get; } 
+
+        /// <summary>
+        /// Gets the current selected value in the input radio group.
+        /// </summary>
+        public object? CurrentValue { get; }
+
+        /// <summary>
+        /// Gets the event callback to be invoked when the selected value is changed.
+        /// </summary>
+        public EventCallback<ChangeEventArgs> ChangeEventCallback { get; }
+
+        /// <summary>
+        /// Instantiates a new <see cref="InputRadioContext" />.
+        /// </summary>
+        /// <param name="parentContext">The parent <see cref="InputRadioContext" />.</param>
+        /// <param name="groupName">The name of the input radio group.</param>
+        /// <param name="currentValue">The current selected value in the input radio group.</param>
+        /// <param name="changeEventCallback">The event callback to be invoked when the selected value is changed.</param>
+        public InputRadioContext(
+            InputRadioContext? parentContext,
+            string groupName,
+            object? currentValue,
+            EventCallback<ChangeEventArgs> changeEventCallback)
+        {
+            _parentContext = parentContext;
+
+            GroupName = groupName;
+            CurrentValue = currentValue;
+            ChangeEventCallback = changeEventCallback;
+        }
+
+        /// <summary>
+        /// Finds an <see cref="InputRadioContext"/> in the context's ancestors with the matching <paramref name="groupName"/>.
+        /// </summary>
+        /// <param name="groupName">The group name of the ancestor <see cref="InputRadioContext"/>.</param>
+        /// <returns>The <see cref="InputRadioContext"/>, or <c>null</c> if none was found.</returns>
+        public InputRadioContext? FindContextInAncestors(string groupName)
+            => string.Equals(GroupName, groupName) ? this : _parentContext?.FindContextInAncestors(groupName);
```

Comment:
👍 This is an extremely neat way of dealing with the nesting and flowing events up the hierarchy.